docs/v1-proposed-api.md

```diff
@@ -0,0 +1,254 @@
+# Proposed Changes for Hive V1 API
+
+Minimal changes, no new CRDs or relationships between.
+
+## ClusterDeployment
+
+### Current V1Alpha1 ClusterDeployment Example
+
+```yaml
+apiVersion: hive.openshift.io/v1alpha1
+kind: ClusterDeployment
+metadata:
+  name: ci-cluster-v4-1
+  namespace: uhc-staging-18urm2sbab85d9crlk27is39n706ph3g
+spec:
+  baseDomain: r4f1.s1.devshift.org
+  certificateBundles:
+  - generate: true
+    name: primary-cert-bundle
+    secretRef:
+      name: primary-cert-bundle-secret
+  clusterName: ci-cluster-v4-1
+  compute:
+  - name: worker
+    platform:
+      aws:
+        rootVolume:
+          iops: 100
+          size: 300
+          type: gp2
+        type: m5.xlarge
+        zones:
+        - us-east-1a
+    replicas: 4
+  controlPlane:
+    name: master
+    platform:
+      aws:
+        rootVolume:
+          iops: 100
+          size: 32
+          type: gp2
+        type: m5.xlarge
+        zones:
+        - us-east-1a
+    replicas: 3
+  controlPlaneConfig:
+    servingCertificates:
+      default: primary-cert-bundle
+  imageSet:
+    name: openshift-v4.1.18
+  images: {}
+  ingress:
+  - domain: apps.ci-cluster-v4-1.r4f1.s1.devshift.org
+    name: default
+    servingCertificate: primary-cert-bundle
+  installed: true
+  manageDNS: true
+  networking:
+    clusterNetworks:
+    - cidr: 10.128.0.0/14
+      hostSubnetLength: 23
+    machineCIDR: 10.0.0.0/16
+    serviceCIDR: 172.30.0.0/16
+    type: OpenShiftSDN
+  platform:
+    aws:
+      region: us-east-1
+  platformSecrets:
+    aws:
+      credentials:
+        name: aws
+  pullSecret:
+    name: pull
+  sshKey:
+    name: ssh
+status:
+  adminKubeconfigSecret:
+    name: ci-cluster-v4-1-0-ndv7l-admin-kubeconfig
+  adminPasswordSecret:
+    name: ci-cluster-v4-1-0-ndv7l-admin-password
+  apiURL: https://api.ci-cluster-v4-1.r4f1.s1.devshift.org:6443
+  certificateBundles:
+  - generated: true
+    name: primary-cert-bundle
+  cliImage: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:528f2ead3d1605bdf818579976d97df5dd86df0a2a5d80df9aa8209c82333a86
+  clusterID: 1e4f00a3-a30e-46d7-9567-4bb61caf6d73
+  clusterVersionStatus:
+    [SNIP]
+  conditions:
+  - lastProbeTime: "2019-10-23T08:40:24Z"
+    lastTransitionTime: "2019-10-23T08:40:24Z"
+    message: Control plane certificates are present
+    reason: ControlPlaneCertificatesFound
+  installed: true
+  installedTimestamp: "2019-10-23T08:38:03Z"
+  installerImage: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:c9c1e6d1897b74d366d331f5cdc3fa16c5a96abe28ff8238833c09042156e983
+  provision:
+    name: ci-cluster-v4-1-0-ndv7l
+  webConsoleURL: https://console-openshift-console.apps.ci-cluster-v4-1.r4f1.s1.devshift.org
+```
+
+### Proposed Changes
+
+
+ 1. Add Spec.Installation struct to contain install specific, immutable settings.
+ 1. Add Spec.Installation.InstallConfigSecret containing a full openshift-install InstallConfig  to pass directly to installer.
```

Comment:
This would be great for flexibility, but not so great for simplicity. It means that the user of the API is no longer isolated from the details of the installer. Is the schema of this installer configuration versioned and backwards compatible? Will this be optional or mandatory?

Reply:
Yes that is the tradeoff, it is a little more complex for callers of Hive's API. You will need to know about ClusterDeployment, as well as InstallConfig. The installer schema is versioned and backwards compatible, but not forwards compatible. Thus far Hive has only needed to know about one version, however it is very likely in the future, you will need to know about new versions of InstallConfig to access new feature, as well as old versions to still be able to install those. This problem was on the horizon for Hive regardless if we were not undertaking this change. It's not going to be fun either if you're trying to do it with Go vendoring.

This will be mandatory, we can't make it optional as part of the API goal here is to eliminate things we don't need, but the InstallConfig does, so we could not generate it from an ClusterDeployment given the proposed changes.

Reply:
Is it possible hive could translate between versions to make it easier for clients?  I.e. version A is what we use now, installer moves to version B.  Hive translates A to B if it sees that version.  Risk is carrying translation logic for a long time, benefit is breaking changes in installer config won't break clients immediately.